### ISSUE TITLE
Add UnliftedRef

### ISSUE THREAD

**Issue by treeowl** [haskell/primitive#198]
I believe we should be able to `unsafeCoerce#` our way into references containing `TYPE 'UnliftedRep`. This is horrible, of course, but also useful. We can support all the usual `MutVar` operations except `atomicModifyMutVar*` ones.

**Comment by andrewthad** [comment]
I don’t quite understand what you mean. Would this just be backed by an UnliftedArray of length 1?

**Comment by andrewthad** [comment]
Ah, thanks for clarifying. Now I get it. This seems reasonable, and it should work. It would be nice to fix `RuntimeRep` as you suggest. The documentation of `mkWeak#` even hints at this shortcoming of the current system.

**Comment by treeowl** [comment]
OTOH, maybe that doesn't help much. We still couldn't write user functions that didn't care about levity....

**Comment by andrewthad** [comment]
True. We could unify the types but we would still need an extra copy of all the functions. Unless someone finds a good way to [allow levity polymorphism in binding positions](https://ghc.haskell.org/trac/ghc/ticket/14917).

**Comment by konsumlamm** [comment]
Since GHC 9.4, `MutVar#` is levity polymorphic, so this can be now implemented by just making `MutVar` levity polymorphic too - no `unsafeCoerce#` needed. See #356.
